### numpy random 함수
- np.random.seed: seed를 통한 난수 생성
- np.random.randint: 정수 난수 생성
- np.random.rand: 0-1까지 균일분포에서 난수 매트릭스 배열 생성
- np.random.randn: 가우시안 표준 정규분포에서 난수 매트릭스 배열 생성
- np.random.shuffle: 기존 데이터의 순서 바꾸기
- np.random.choice: 기존 데이터에서 sampling

#### random 라이브러리

In [2]:
import random

In [6]:
print(random.random(),'\n')            #0-1 사이 랜덤 float 추출

print(random.randint(1,10),'\n')       #1-10 사이 랜덤 int추출

print(random.randrange(1,10),'\n')       #0-(지정범위-1) 랜덤 int 추출

print(random.uniform(10,20),'\n')      #10-20 사이 랜덤 float 추출


0.3045167866849232 

8 

9 

11.228507945462853 



In [465]:
li=list(range(20))+[1,2,3]

print(random.choice(li))       #리스트 내 랜덤 추출

print(random.sample(li,3))     #리스트 내 랜덤으로 중복 없이 n개 추출

random.shuffle(li)             #리스트 데이터 섞어줌, 리스트를 따로 출력해서 알아볼 수 있음
print(li)


3
[9, 4, 14]
[7, 12, 13, 11, 3, 5, 15, 16, 0, 2, 3, 6, 10, 14, 8, 1, 4, 19, 9, 2, 18, 17, 1]


#### randint
- 파이썬 random 라이브러리 내 는 끝 구간을 포함하는데, numpy에서는 끝 구간을 포함하지 않음

In [290]:
np.random.randint(1,10,10)

array([5, 8, 5, 2, 1, 9, 2, 5, 2, 1])

#### rand
- [0,1] 균일분포 내 난수 생성

In [13]:
import numpy as np

In [14]:
r=np.random.rand(3,3)
r

array([[0.20561845, 0.76627777, 0.8648064 ],
       [0.77304048, 0.28150533, 0.46406015],
       [0.38542755, 0.67734352, 0.81853972]])

In [15]:
#평균 알아보기

r.mean()     #0.5에 가까움

0.5818465973523529

In [16]:
#난수 개수 늘리기

r2=np.random.rand(10000,10000)

In [17]:
r2.mean()     #0.5에 한층 더 가까워짐

0.4999754190022042

#### randn
- 표준정규분포 내 난수 생성

In [18]:
r=np.random.randn(7)
r

array([ 0.33459453,  0.05238293,  0.14700894,  0.73272876,  0.44684252,
       -0.5373432 , -1.32960097])

In [19]:
#평균 알아보기

r.mean()

-0.021912357439344245

In [20]:
#난수 개수를 늘리기

r2=np.random.randn(1000000)

In [23]:
r2.mean()  #평균이 0에 가까워짐
'%.5f'%r2.mean()

'-0.00117'

#### uniform
- 지정 구간 내 float 난수 생성

In [24]:
r=np.random.uniform(4,5,size=(5,5))
r

array([[4.94555322, 4.67033371, 4.86792041, 4.56015923, 4.95790005],
       [4.84743068, 4.95641842, 4.14163155, 4.06128598, 4.05899398],
       [4.29386029, 4.78359568, 4.06866064, 4.19778525, 4.47693672],
       [4.9315742 , 4.34800986, 4.24134645, 4.04069028, 4.62788495],
       [4.30540344, 4.96775531, 4.03108299, 4.37537362, 4.11678262]])

In [25]:
r=np.random.uniform(1,2,(3,3))
r

array([[1.21562156, 1.34471199, 1.94841984],
       [1.52466068, 1.85440374, 1.83516706],
       [1.61684375, 1.7453046 , 1.56223805]])

#### random_sample
- 0-1 구간 내 난수 생성
- uniform 함수의 구간을 0,1로 지정한것과 같음
- random 함수와 같음

In [198]:
np.random.random_sample((5,5))

array([[0.37867345, 0.45674386, 0.16568636, 0.68565481, 0.39645794],
       [0.82256637, 0.07313181, 0.3813022 , 0.04255891, 0.62383639],
       [0.50505959, 0.49454356, 0.87647111, 0.73174829, 0.03098709],
       [0.21648283, 0.2726576 , 0.86315884, 0.29331338, 0.13411857],
       [0.86098261, 0.2844522 , 0.98554799, 0.22053714, 0.95401649]])

In [193]:
np.random.random((5,5))

array([[0.68878665, 0.13786821, 0.10883145, 0.45830322, 0.57317737],
       [0.41215315, 0.05470162, 0.22319959, 0.69291803, 0.16763301],
       [0.4708954 , 0.78783358, 0.55134363, 0.42094179, 0.7644009 ],
       [0.93223014, 0.75901273, 0.98409489, 0.53570853, 0.32360154],
       [0.32183428, 0.97607016, 0.14971885, 0.96309668, 0.10179592]])

In [199]:
np.random.uniform(0,1,(5,5))

array([[0.20186562, 0.86376593, 0.57670727, 0.10329041, 0.4105371 ],
       [0.15624883, 0.26776169, 0.62871956, 0.52117574, 0.49414941],
       [0.36637457, 0.67214063, 0.70482601, 0.60739515, 0.85362103],
       [0.26516287, 0.49750635, 0.28859845, 0.90926721, 0.94743571],
       [0.94053747, 0.10965041, 0.53107031, 0.91892917, 0.21117213]])

#### 실습

In [212]:
import pandas as pd

In [26]:
ar=np.random.random_sample((8,8))
ar

array([[0.0476233 , 0.62017686, 0.38509564, 0.20840971, 0.1771956 ,
        0.28963432, 0.96730855, 0.87853694],
       [0.20255691, 0.1794864 , 0.94129588, 0.16067389, 0.47878717,
        0.84567841, 0.14982646, 0.97666973],
       [0.7802154 , 0.15701602, 0.68515368, 0.19024897, 0.66196011,
        0.1784976 , 0.55517745, 0.52669605],
       [0.6447858 , 0.34588943, 0.6628911 , 0.73008348, 0.96057648,
        0.58614247, 0.09558851, 0.95442209],
       [0.19749288, 0.87323435, 0.23475711, 0.91697159, 0.97120414,
        0.5107577 , 0.30857826, 0.18383766],
       [0.33168442, 0.77924113, 0.64526511, 0.27250492, 0.70884444,
        0.18358098, 0.02707767, 0.84211027],
       [0.2134772 , 0.73511377, 0.54093243, 0.48025144, 0.08297487,
        0.1270501 , 0.12105139, 0.76361749],
       [0.39536851, 0.43236378, 0.66885613, 0.56927374, 0.39376159,
        0.00970754, 0.44285545, 0.62643667]])

In [228]:
#생성된 난수들로 데이터프레임 만들기

df=pd.DataFrame(ar,columns=['id','gender','age','region',
                           'product','price','qty','time'])
df

id    gender       age    region   product     price       qty  \
0  0.257137  0.147867  0.358871  0.276558  0.537922  0.175251  0.929622   
1  0.598527  0.963408  0.199136  0.770164  0.880941  0.865397  0.270121   
2  0.905512  0.027777  0.201325  0.151019  0.771609  0.889105  0.523992   
3  0.925869  0.769423  0.744661  0.718995  0.563445  0.104435  0.255231   
4  0.988649  0.947582  0.404001  0.739200  0.147398  0.881623  0.014951   
5  0.458763  0.250081  0.100104  0.164503  0.611119  0.839969  0.938163   
6  0.741441  0.032725  0.626340  0.314652  0.301880  0.010815  0.452675   
7  0.398605  0.045936  0.504287  0.315209  0.952258  0.702382  0.465140   

       time  
0  0.035585  
1  0.139073  
2  0.509250  
3  0.996359  
4  0.428737  
5  0.576881  
6  0.660005  
7  0.862764

In [229]:
#id, gender, age를 의미 있는 데이터로 바꾸기

#id 고유화하기
from sklearn.preprocessing import LabelEncoder

le=LabelEncoder()
df['id']=le.fit_transform(df['id'])
df=df.sort_values('id',ignore_index=True)  # 정렬 순서대로 인덱스 세팅 다시~


#gender 삼항식으로 이분화하기
df['gender']=np.where(df['gender']>0.5,'male','female')


#age 나이 계산하기
df['age']=round(df['age']*100)

df3=df.iloc[:,:3]
df3.head()

id  gender   age
0   0  female  36.0
1   1  female  50.0
2   2  female  10.0
3   3    male  20.0
4   4  female  63.0

In [243]:
df3.gender.value_counts()

female    5
male      3
Name: gender, dtype: int64

In [242]:
df3.gender.unique()

array(['female', 'male'], dtype=object)

#### 실습2

In [230]:
#배열 정렬 후, 상위 50% 원소 추출

ar=np.random.randn(20)
ar

array([-0.35408288, -1.29815075, -0.27534205, -0.93603391,  1.88363041,
       -0.02953076, -0.32595228,  2.16634053, -1.24673995, -2.35607717,
       -1.50046226,  0.05228149,  0.85280106, -0.9148975 , -0.17408592,
       -0.66648111, -0.51138476, -0.56004803,  0.91883903, -0.70079521])

In [233]:
ar.sort()
ar

array([-2.35607717, -1.50046226, -1.29815075, -1.24673995, -0.93603391,
       -0.9148975 , -0.70079521, -0.66648111, -0.56004803, -0.51138476,
       -0.35408288, -0.32595228, -0.27534205, -0.17408592, -0.02953076,
        0.05228149,  0.85280106,  0.91883903,  1.88363041,  2.16634053])

In [236]:
ar[int(len(ar)*0.5)]       #0.5를 곱하면서 실수가 되므로, 정수화가 필요함

-0.3540828760065148

#### 실습3

In [291]:
#배열의 집합 연산

names=np.array(['Joe','Bob','Andrew','Joe','Joe'])

In [293]:
np.unique(names)  #원소 value 확인

array(['Andrew', 'Bob', 'Joe'], dtype='<U6')

In [301]:
sorted(set(names))   #sorted 내장함수를 사용함으로써 list화 됨

['Andrew', 'Bob', 'Joe']

#### save, load
- 데이터 저장,출력(입출력)

In [302]:
ar=np.arange(10)
ar

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [303]:
np.save('저장한닷',ar)     #저장

In [304]:
np.load('저장한닷.npy')     #불러오기

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

#### savez
- 여러개 배열을 압축된 형식으로 저장

In [311]:
np.savez('저장한닷닷',a=ar,b=ar*2,c=ar*3) #저장

In [316]:
load=np.load('저장한닷닷.npz')
load['a'], load['b'], load['c']

(array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
 array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18]),
 array([ 0,  3,  6,  9, 12, 15, 18, 21, 24, 27]))

#### argsort
- 정렬된 데이터들의 원본 인덱스 주소를 출력

In [318]:
ar=np.random.randint(10,50,30)
ar

array([25, 35, 33, 21, 22, 21, 25, 34, 30, 15, 26, 21, 46, 12, 25, 38, 39,
       21, 32, 25, 44, 11, 47, 33, 13, 47, 45, 26, 31, 12])

In [319]:
np.argsort(ar)

array([21, 29, 13, 24,  9,  3,  5, 17, 11,  4, 19,  0, 14,  6, 10, 27,  8,
       28, 18,  2, 23,  7,  1, 15, 16, 20, 26, 12, 22, 25], dtype=int64)

In [322]:
#역으로 정렬된 인덱스 주소 출력

np.argsort(ar)[::-1]

array([25, 22, 12, 26, 20, 16, 15,  1,  7, 23,  2, 18, 28,  8, 27, 10,  6,
       14,  0, 19,  4, 11, 17,  5,  3,  9, 24, 13, 29, 21], dtype=int64)

#### 실습

In [326]:
#0-10 정수(5,5) 행렬의 역행렬 구하고, 두 행렬간 곱 구하기

rand=np.random.randint(0,11,(5,5))
print(rand)

rand_inv=np.linalg.inv(rand)
print(rand_inv)

np.dot(rand,rand_inv).round()

[[10 10  4  8  4]
 [ 3  2  6  9  7]
 [ 9  0  6  0  2]
 [ 9  0  2  7  4]
 [ 6  4  9  8  2]]
[[ 0.01790963 -0.06247864  0.04573108  0.07532982 -0.01353476]
 [ 0.10940597  0.00382801 -0.00499009 -0.10089548 -0.02542894]
 [-0.03937385  0.03048739  0.06740037 -0.08927473  0.08319092]
 [-0.03322168 -0.03677627 -0.13063094  0.11217445  0.10144234]
 [ 0.0375282   0.18969171  0.09200902 -0.07116002 -0.18866635]]


array([[ 1.,  0.,  0., -0.,  0.],
       [ 0.,  1., -0., -0., -0.],
       [-0.,  0.,  1.,  0.,  0.],
       [-0., -0., -0.,  1.,  0.],
       [ 0.,  0., -0., -0.,  1.]])

### 도전과제(7/6)
- np.arange(336).reshape(6,7,8)에서 100번째 요소의 인덱스 구하기

In [334]:
ar=np.arange(336).reshape(6,7,8)
ar[0]

array([[ 0,  1,  2,  3,  4,  5,  6,  7],
       [ 8,  9, 10, 11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20, 21, 22, 23],
       [24, 25, 26, 27, 28, 29, 30, 31],
       [32, 33, 34, 35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44, 45, 46, 47],
       [48, 49, 50, 51, 52, 53, 54, 55]])

In [330]:
for i in range(6):
    for j in range(7):
        for k in range(8):
            if ar[i,j,k]==100:
                print(f'원소 100의 인덱스 번호는 ({i},{j},{k})입니다.')

원소 100의 인덱스 번호는 (1,5,4)입니다.


In [331]:
#확인하기
ar[1,5,4]

100

### 과제2
아래 가이드에 따라 고객별 연간 구매데이터를 생성하고 아래사항을 수행하기(관측치 1,000개)

#### Q1. 고객 속성(id, 성별, 연령, 거주지역) 데이터와 거래 데이터를 통합한 데이터 프레임 df를 생성하세요.

거래 번호
- id : 1 ~ 1000번 일련번호 - 개시 후 1천 건의 거래가 있었고 걍 거기에 넘버링한 거라고 이해해보자

고객별 속성 데이터
- cusno : 고객번호 1000개 랜덤
- gender : 0,1 정수 난수 생성
- age : 0 ~ 80사이 정수 난수 생성
- region : 1 ~ 10 사이 정수 난수 생성

거래 데이터
- (day : 요일)
- (date : 2020-01-01 ~ 2021-12-31 사이의 날짜 1000개 생성)
- product_20 : 고가제품(10), 중가제품(20), 저가제품(30)로 구분제품코드 10,20,30
- product_21 : 고가제품(10), 중가제품(20), 저가제품(30)로 구분제품코드 10,20,30
- price_20 : 1000 ~ 50000 사이 실수 난수 생성
- price_21 : 1000 ~ 50000 사이 실수 난수 생성
- quantity_20 : 1 ~ 100 사이 정수 난수 생성
- quantity_21 : 1 ~ 100 사이 정수 난수 생성
- (time_20 : 01 ~ 24 사이 정수 난수 생성)
- (time_21 : 01 ~ 24 사이 정수 난수 생성)
- amount_20 : price_20 * quantity_20
- amount_21 : price_21 * quantity_21
- sales : 20년 대비 21년구매금액이 증가면 1, 감소면 0



※주의할 점
- date는 product20,21에 주의해서 작성
    - day는 date의 요일
    - time_20,21는 date의 시간

In [30]:
import numpy as np
import pandas as pd
import random

np.random.seed(100)

#고객 데이터프레임 생성
id=pd.Series(np.arange(1,1001))
cusno=pd.Series(np.random.randint(1000,5000,size=1000))
gender=pd.Series(np.random.randint(0,2,size=1000))
age=pd.Series(np.random.randint(0,80,size=1000))
region=pd.Series(np.random.randint(1,11,size=1000))

customer=pd.concat([id,cusno,gender,age,region],axis=1)
customer.rename(columns={0:'id',1:'cusno',2:'gender',3:'age',4:'region'},
               inplace=True)


#구매 데이터프레임 생성
product_20=pd.Series(np.random.choice([10,20,30],size=1000))
product_21=pd.Series(np.random.choice([10,20,30],size=1000))
price_20=pd.Series(np.random.uniform(1000,50000,size=1000).round(2))
price_21=pd.Series(np.random.uniform(1000,50000,size=1000).round(2))
quantity_20=pd.Series(np.random.randint(1,100,size=1000) )
quantity_21=pd.Series(np.random.randint(1,100,size=1000))
amount_20=pd.Series(np.array(price_20*quantity_20).round(2))
amount_21=pd.Series(np.array(price_21*quantity_21).round(2))
sales=pd.Series(np.where(amount_20<amount_21,1,0))

purchase=pd.concat([product_20,product_21,price_20,
                    quantity_20,quantity_21,amount_20,
                    amount_21,sales],axis=1)
purchase.rename(columns={0:'product_20',
                         1:'product_21',
                         2:'price_20',
                         3:'price_21',
                         4:'quantity_20',
                         5:'quantity_21',
                         6:'amount_20',
                         7:'amount_21',
                         8:'sales'},
               inplace=True)


#데이터프레임 합치기
df=customer.join(purchase,how='outer')
df.head()

id  cusno  gender  age  region  product_20  product_21  price_20  price_21  \
0   1   2544       1   48       1          20          10  23225.68        96   
1   2   3840       0   20       8          10          20  27716.36        92   
2   3   2859       0   54       7          10          10  24624.63         7   
3   4   4943       1   79       1          30          20  44141.70        51   
4   5   4927       0   10       3          30          20  26712.91        68   

   quantity_20  quantity_21   amount_20  amount_21  
0           91   2229665.28  2183282.01          0  
1            5   2549905.12    52445.30          0  
2           92    172372.41  2348875.00          1  
3           20   2251226.70   204012.80          0  
4           38   1816477.88   590478.20          0

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           1000 non-null   int32  
 1   cusno        1000 non-null   int32  
 2   gender       1000 non-null   int32  
 3   age          1000 non-null   int32  
 4   region       1000 non-null   int32  
 5   product_20   1000 non-null   int32  
 6   product_21   1000 non-null   int32  
 7   price_20     1000 non-null   float64
 8   price_21     1000 non-null   int32  
 9   quantity_20  1000 non-null   int32  
 10  quantity_21  1000 non-null   float64
 11  amount_20    1000 non-null   float64
 12  amount_21    1000 non-null   int32  
dtypes: float64(3), int32(10)
memory usage: 62.6 KB


In [32]:
df['cusno'].duplicated().sum()

118

#### Q2.data 유의미하게 5가지 도출

1. 코로나로 인해, 21년도에 새로 유입된 소비자는 20년도보다 10%p 상승
    - product_20 데이터 90개를 np.nan화,cusno 200개 중복 x
2. 21년도 이탈 고객은 주로 남성들
    - product_20가 np.nan이 아니고, product_21 데이터 90개를 np.nan화, gender을 거의 남성으로 전환
3. 코로나로 인해, 생필품 구비율이 증가해 20년도 대비 21년도 중저가 소비가 증가함
    - product_21이 10,20이면, quantity_21 증가
4. 재구매가 자주 이루어진 고객은 보통 여성들
    - cusno의 랜덤 범위를 줄이고, 중복된 cusno의 gender을 거의 여성으로 전환
5. 

##### 1. 20년도, 21년도 재구매가 이루어진 고객수의 80%가 여성

In [33]:
#재구매 횟수
pd.set_option('display.max_row',120)

re_purchase=df['cusno'].duplicated()
print(df.loc[re_purchase,'id'].value_counts())
print(df.loc[re_purchase,'id'].value_counts().unique())


142    1
796    1
871    1
866    1
864    1
853    1
846    1
844    1
839    1
837    1
826    1
820    1
818    1
804    1
788    1
875    1
787    1
783    1
777    1
769    1
767    1
760    1
759    1
755    1
751    1
743    1
741    1
736    1
872    1
877    1
158    1
950    1
987    1
981    1
980    1
975    1
974    1
971    1
969    1
964    1
959    1
953    1
952    1
951    1
949    1
888    1
942    1
936    1
935    1
932    1
929    1
917    1
915    1
914    1
910    1
905    1
901    1
896    1
733    1
725    1
720    1
387    1
541    1
540    1
531    1
523    1
515    1
513    1
485    1
481    1
446    1
440    1
430    1
398    1
341    1
715    1
340    1
324    1
319    1
311    1
301    1
298    1
291    1
286    1
266    1
255    1
243    1
228    1
549    1
555    1
561    1
563    1
707    1
705    1
700    1
699    1
698    1
695    1
691    1
685    1
684    1
680    1
676    1
672    1
661    1
660    1
652    1
641    1
636    1
634    1
632    1
6

In [34]:
#재구매 고객

before_all=len(df[re_purchase])   #재구매 고객수
print('총 재구매 고객 수: ',before_all)

print('재구매 고객의 80%: ',int(before_all*0.8))
pd.set_option('display.max_row',120)

df.loc[re_purchase,'id'].value_counts().unique()
before_f=len(df.loc[(re_purchase)&(df['gender']==1),'gender']) #그 중 여성의 수
before_m=len(df.loc[(re_purchase)&(df['gender']==0),'gender'])
print('재구매 고객 중 여성 수: ',before_f)
print('재구매 고객 중 남성 수: ',before_m)

총 재구매 고객 수:  118
재구매 고객의 80%:  94
재구매 고객 중 여성 수:  59
재구매 고객 중 남성 수:  59


In [35]:
#재구매 남성 인덱스 사용하여 36개를 랜덤 추출 
rep_g_idx=df.loc[(re_purchase)&(df['gender']==0)].index.tolist()
rep_g_idx=random.sample(rep_g_idx,35)

print(rep_g_idx)
print(len(rep_g_idx))

[900, 719, 714, 776, 323, 290, 825, 740, 874, 974, 803, 951, 540, 560, 735, 754, 539, 950, 651, 671, 683, 697, 227, 871, 863, 913, 562, 631, 635, 732, 931, 618, 310, 980, 750]
35


In [36]:
#인덱스를 활용하여 여성으로 변환

df.loc[rep_g_idx,'gender']=1

after_f=len(df[(re_purchase)&(df['gender']==1)])
after_m=len(df[(re_purchase)&(df['gender']==0)])
print('변환 후 재구매 여성 고객 수: ',after_f)
print('변환 후 재구매 남성 고객 수: ',after_m)

변환 후 재구매 여성 고객 수:  94
변환 후 재구매 남성 고객 수:  24


In [37]:
#재구매여부 변수를 데이터프레임에 추가

df['re_purchase']=re_purchase.replace([False,True],[0,1])
print(df['re_purchase'])

0      0
1      0
2      0
3      0
4      0
      ..
995    0
996    0
997    0
998    0
999    0
Name: re_purchase, Length: 1000, dtype: int64


In [38]:
df.head()

id  cusno  gender  age  region  product_20  product_21  price_20  price_21  \
0   1   2544       1   48       1          20          10  23225.68        96   
1   2   3840       0   20       8          10          20  27716.36        92   
2   3   2859       0   54       7          10          10  24624.63         7   
3   4   4943       1   79       1          30          20  44141.70        51   
4   5   4927       0   10       3          30          20  26712.91        68   

   quantity_20  quantity_21   amount_20  amount_21  re_purchase  
0           91   2229665.28  2183282.01          0            0  
1            5   2549905.12    52445.30          0            0  
2           92    172372.41  2348875.00          1            0  
3           20   2251226.70   204012.80          0            0  
4           38   1816477.88   590478.20          0            0

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           1000 non-null   int32  
 1   cusno        1000 non-null   int32  
 2   gender       1000 non-null   int32  
 3   age          1000 non-null   int32  
 4   region       1000 non-null   int32  
 5   product_20   1000 non-null   int32  
 6   product_21   1000 non-null   int32  
 7   price_20     1000 non-null   float64
 8   price_21     1000 non-null   int32  
 9   quantity_20  1000 non-null   int32  
 10  quantity_21  1000 non-null   float64
 11  amount_20    1000 non-null   float64
 12  amount_21    1000 non-null   int32  
 13  re_purchase  1000 non-null   int64  
dtypes: float64(3), int32(10), int64(1)
memory usage: 70.4 KB


##### 2. 20년도에 구매가 이루어졌으나, 21년도 구매 이력이 없는 이탈 고객 수의 80%가 남자
- 조건: 재구매 고객 제외

In [40]:
#재구매 고객이 아닌 고객 확인

n_all_re_purchase=len(df[(re_purchase==False)])
print('재구매 고객이 아닌 총 고객수: ',n_all_re_purchase)

재구매 고객이 아닌 총 고객수:  882


In [41]:
#현재 데이터 내 21년도 구매이력이 없는 고객 수 확인

before_m=df.loc[(re_purchase==False)&(df['gender']==0),
                ['product_21']].isnull().sum()
before_f=df.loc[(re_purchase==False)&(df['gender']==1),
                ['product_21']].isnull().sum()

print('이탈 남자 고객 수:\n',before_m)
print('이탈 여자 고객 수:\n',before_f)

이탈 남자 고객 수:
 product_21    0
dtype: int64
이탈 여자 고객 수:
 product_21    0
dtype: int64


In [42]:
#21년도 구매이력 없는 90명의 이탈고객 만들기

print('이탈 해야할 남자 고객 수: ',int(90*0.8),'명')
print('이탈 해야할 여자 고객 수: ',int(90*0.2),'명')

이탈 해야할 남자 고객 수:  72 명
이탈 해야할 여자 고객 수:  18 명


In [43]:
#90명의 이탈 고객: 남자:72명, 여자:12명의 인덱스 랜덤 추출
bye_m_idx=df.loc[(re_purchase==False)&(df['gender']==0)].index.tolist()
bye_m_idx=random.sample(bye_m_idx,72)

bye_f_idx=df.loc[(re_purchase==False)&(df['gender']==1)].index.tolist()
bye_f_idx=random.sample(bye_f_idx,18)

In [44]:
#추출된 인덱스로 21년도 구매이력 지우기 
df.loc[bye_m_idx,['product_21','price_21','quantity_21','amount_21']]=np.nan
df.loc[bye_f_idx,['product_21','price_21','quantity_21','amount_21']]=np.nan

after_m=df.loc[(re_purchase==False)&(df['gender']==0),
                ['product_21']].isnull().sum()
after_f=df.loc[(re_purchase==False)&(df['gender']==1),
                ['product_21']].isnull().sum()

print('이탈한 남자 고객 수:\n',after_m)
print('이탈한 여자 고객 수:\n',after_f)

이탈한 남자 고객 수:
 product_21    72
dtype: int64
이탈한 여자 고객 수:
 product_21    18
dtype: int64


In [45]:
#이탈 고객 여부 변수 생성

df['bye_purchase']=np.where(df['product_21'].isnull(),1,0)
df['bye_purchase'].value_counts()

0    910
1     90
Name: bye_purchase, dtype: int64

#### Q3. df 를 탐색적 분석을 통하여 인사이트를 도출하세요.

In [46]:
df.corr()

id     cusno    gender       age    region  product_20  \
id            1.000000 -0.033492  0.044722  0.010530  0.010237   -0.021646   
cusno        -0.033492  1.000000 -0.031681 -0.018124  0.045879    0.050931   
gender        0.044722 -0.031681  1.000000 -0.027116 -0.001476    0.039173   
age           0.010530 -0.018124 -0.027116  1.000000 -0.046480    0.007771   
region        0.010237  0.045879 -0.001476 -0.046480  1.000000   -0.065325   
product_20   -0.021646  0.050931  0.039173  0.007771 -0.065325    1.000000   
product_21    0.038882  0.015965 -0.027668 -0.057740  0.034207    0.018110   
price_20     -0.032281 -0.003011  0.019647 -0.028799 -0.019255    0.018662   
price_21      0.014808  0.021398  0.009751  0.019117 -0.055529    0.051633   
quantity_20  -0.018543  0.023929  0.013351  0.008729  0.008131    0.000444   
quantity_21  -0.025738  0.001942  0.026344 -0.029064 -0.050789    0.028182   
amount_20     0.002609  0.062306  0.047681  0.026291  0.020932   -0.036980   
amount_21    -0.003711  0.055870  0.042548 -0.001762  0.024129   -0.033714   
re_purchase   0.243007 -0.043672  0.215520 -0.022512  0.024482   -0.055099   
bye_purchase -0.004721 -0.012586 -0.189951 -0.023282  0.058905    0.005815   

              product_21  price_20  price_21  quantity_20  quantity_21  \
id              0.038882 -0.032281  0.014808    -0.018543    -0.025738   
cusno           0.015965 -0.003011  0.021398     0.023929     0.001942   
gender         -0.027668  0.019647  0.009751     0.013351     0.026344   
age            -0.057740 -0.028799  0.019117     0.008729    -0.029064   
region          0.034207 -0.019255 -0.055529     0.008131    -0.050789   
product_20      0.018110  0.018662  0.051633     0.000444     0.028182   
product_21      1.000000 -0.026793  0.024106    -0.031712    -0.004420   
price_20       -0.026793  1.000000 -0.012461    -0.002285     0.651844   
price_21        0.024106 -0.012461  1.000000     0.011452     0.648802   
quantity_20    -0.031712 -0.002285  0.011452     1.000000     0.002630   
quantity_21    -0.004420  0.651844  0.648802     0.002630     1.000000   
amount_20      -0.007087 -0.054399 -0.050820     0.667178    -0.068767   
amount_21      -0.002529 -0.395212 -0.443529     0.362150    -0.585611   
re_purchase    -0.009147  0.002984  0.014290    -0.029415     0.008153   
bye_purchase         NaN  0.008108       NaN     0.033992          NaN   

              amount_20  amount_21  re_purchase  bye_purchase  
id             0.002609  -0.003711     0.243007     -0.004721  
cusno          0.062306   0.055870    -0.043672     -0.012586  
gender         0.047681   0.042548     0.215520     -0.189951  
age            0.026291  -0.001762    -0.022512     -0.023282  
region         0.020932   0.024129     0.024482      0.058905  
product_20    -0.036980  -0.033714    -0.055099      0.005815  
product_21    -0.007087  -0.002529    -0.009147           NaN  
price_20      -0.054399  -0.395212     0.002984      0.008108  
price_21      -0.050820  -0.443529     0.014290           NaN  
quantity_20    0.667178   0.362150    -0.029415      0.033992  
quantity_21   -0.068767  -0.585611     0.008153           NaN  
amount_20      1.000000   0.571251    -0.028865      0.011218  
amount_21      0.571251   1.000000    -0.028471           NaN  
re_purchase   -0.028865  -0.028471     1.000000     -0.115029  
bye_purchase   0.011218        NaN    -0.115029      1.000000

In [47]:
df.corr().loc[:,['gender']].round(2)

gender
id              0.04
cusno          -0.03
gender          1.00
age            -0.03
region         -0.00
product_20      0.04
product_21     -0.03
price_20        0.02
price_21        0.01
quantity_20     0.01
quantity_21     0.03
amount_20       0.05
amount_21       0.04
re_purchase     0.22
bye_purchase   -0.19

gender와 다른 변수들 간의 상관관계가 절대값 0.05 미만으로 연관성이 낮게 나타난 것과 달리,
re_purchase(재구매 여부)의 상관관계 계수는 0.22, bye_purchase(21년도 구매 이탈 여부)의 상관계수는 -0.19로 어느정도 상관관계를 유추할 수 있는 값이 나옴

따라서, 성별과 재구매여부, 성별과 이탈 고객 여부 등 범주형 자료의 탐색적 분석이 필요함